Code/Framework/AzCore/AzCore/Metrics/EventLoggerFactoryImpl.cpp

```diff
@@ -0,0 +1,145 @@
+        if (loggerIdIter != m_eventLoggerIndirectSet.end()
+            && loggerIdIter->m_index < m_eventLoggers.size())
+        {
+            // The event logger has been found using the event logger id,
+            // so second registration cannot be done
+            return AZ::Failure(AZStd::move(eventLoggerPtr));
+        }
+
+        // Insert new interface since it is not registered
+        m_eventLoggers.emplace_back(AZStd::move(eventLoggerPtr));
+        const size_t emplaceIndex = m_eventLoggers.size() - 1;
+
+        // Use UpperBound to find the insertion slot for the new entry 
+        auto FindIdIndexEntry = [](const EventLoggerIdIndexEntry& lhs, const EventLoggerIdIndexEntry& rhs)
+        {
+            return lhs.m_id < rhs.m_id;
+        };
+        EventLoggerIdIndexEntry newEntry{ eventLoggerId, emplaceIndex };
+        m_eventLoggerIndirectSet.insert(AZStd::upper_bound(m_eventLoggerIndirectSet.begin(), m_eventLoggerIndirectSet.end(),
+            newEntry, AZStd::move(FindIdIndexEntry)),
+            AZStd::move(newEntry));
+
+        return AZ::Success();
+    }
+
+    bool EventLoggerFactoryImpl::UnregisterEventLogger(EventLoggerId eventLoggerId)
+    {
+        AZStd::scoped_lock lock(m_eventLoggerMutex);
+        auto loggerIdIter = FindEventLoggerIndex(eventLoggerId);
+        if (loggerIdIter != m_eventLoggerIndirectSet.end())
+        {
+            // Check if the logger id index is within the event loggers vector
+            if (loggerIdIter->m_index < m_eventLoggers.size())
+            {
+                // Remove the EventLogger
+                m_eventLoggers.erase(m_eventLoggers.begin() + loggerIdIter->m_index);
+            }
+
+            // Remove the Event LoggerId to event logger vector indirection entry as well
+            m_eventLoggerIndirectSet.erase(loggerIdIter);
+            return true;
+        }
+
+        return false;
+    }
+
+    IEventLogger* EventLoggerFactoryImpl::FindEventLogger(EventLoggerId eventLoggerId) const
+    {
+        AZStd::scoped_lock lock(m_eventLoggerMutex);
+        auto loggerIdIter = FindEventLoggerIndex(eventLoggerId);
+        return loggerIdIter != m_eventLoggerIndirectSet.end() && loggerIdIter->m_index < m_eventLoggers.size()
+            ? m_eventLoggers[loggerIdIter->m_index].get() : nullptr;
+    }
+
+    bool EventLoggerFactoryImpl::IsRegistered(EventLoggerId eventLoggerId) const
+    {
+        AZStd::scoped_lock lock(m_eventLoggerMutex);
+        auto loggerIdIter = FindEventLoggerIndex(eventLoggerId);
+        return loggerIdIter != m_eventLoggerIndirectSet.end() && loggerIdIter->m_index < m_eventLoggers.size();
+    }
+
+    // returns iterator containing index into the m_eventLoggers vector that matches the id
+    // otherwise returns the end iteator
+    auto EventLoggerFactoryImpl::FindEventLoggerIndex(EventLoggerId eventLoggerId) const
+        -> typename EventLoggerIndirectSet::const_iterator
```

Comment:
Maybe make a note that this function assumes caller has locked the `m_eventLoggerMutex`.

Reply:
Well this is a private function.

Reply:
I'll will add a comment as well

Reply:
Also: "iteator" -> "iterator"